abdiff/core/collate_ab_transforms.py

```diff
@@ -167,89 +173,90 @@ def get_transformed_batches_iter(
-           Note: This will return the results as a pyarrow.RecordBatchReader.
-
-        4. Yield batches of the joined records until the pyarrow.RecordBatchReader
-           is empty.
-
-    The returned generator can be passed to abdiff.core.utils.write_to_dataset()
-    to perform batch writes of joined TIMDEX A/B records to a Parquet dataset
+    As the records are joined, record batches are yielded which get written to a new,
+    temporary "joined" dataset on disk.
 
     Args:
         dataset_directory: The root directory of the Parquet dataset of TIMDEX records
             (i.e., the tempfile.TemporaryDirectory).
     """
+    partition_paths = glob.glob(f"{dataset_directory}/transformed_file_name=*")
+    transformed_file_names = [os.path.basename(p).split("=")[1] for p in partition_paths]
+
     with duckdb.connect(":memory:") as con:
-        transformed_file_names = con.execute(
-            """
-            SELECT DISTINCT transformed_file_name FROM
-            read_parquet($transformed_parquet_glob, hive_partitioning=true)
-            """,
-            {"transformed_parquet_glob": f"{dataset_directory}/**/*.parquet"},
-        ).fetch_df()["transformed_file_name"]
+        for i, transformed_file in enumerate(transformed_file_names):
+            logger.debug(
+                f"Joining records batch {i+1}/{len(transformed_file_names)}: "
+                f"{transformed_file}"
+            )
+
+            transformed_file_parquet_glob = (
+                f"{dataset_directory}/transformed_file_name={transformed_file}/*.parquet"
+            )
 
-        for transformed_file in transformed_file_names:
             results = con.execute(
                 """
                 WITH
                     transformed_file AS (
                         SELECT * FROM
                         read_parquet(
-                            $transformed_parquet_glob,
+                            $transformed_file_parquet_glob,
                             hive_partitioning=true
                         )
-                        WHERE transformed_file_name=$transformed_file_name
                     ),
                     a AS (SELECT * FROM transformed_file WHERE version='a'),
                     b AS (SELECT * FROM transformed_file WHERE version='b')
                 SELECT
-                    COALESCE(a.timdex_record_id, b.timdex_record_id) timdex_record_id,
-                    COALESCE(a.source, b.source) source,
-                    COALESCE(a.run_date, b.run_date) run_date,
-                    COALESCE(a.run_type, b.run_type) run_type,
-                    COALESCE(a.action, b.action) "action",
-                    a.record as record_a,
-                    b.record as record_b
+                    uuid() as abdiff_record_id,
+                    a.timdex_record_id,
+                    a.source,
+                    a.run_date,
+                    a.run_type,
+                    COALESCE(a.action, b.action) AS action,
```

Comment:
Noting that this `COALESCE(a.action, b.action) AS action` remains non-deterministic.  @jonavellecuerdo pointed out that two different versions of Transmogrifier could _technically_ result in the same file getting interpreted as "to index" or "to delete" which would vary this `action` value.  

It is proposed that we ignore for now.  It remains a very unlikely edge case, and even more so, it actually exposes something that ABDiff is not equipped to handle yet: what if a `timdex_record_id` changes for the same intellectual record?  or in this case, the `action` changes?

This is getting into territory we should explore sometime -- what it means when Transmog varies something fundamental about the record, beyond just the metadata -- but I would propose as out of scope to ABDiff (at this time) merely showing us differences in metadata.